File: sprint8/By languages/go/lesson6code1.go
func Search(p, text string) []int {
    var result []int
    s := p + "#" + text
    π := make([]int, len(p))
    var πPrev int
    for i := 1; i < len(s); i++ {
        k := πPrev
        for k > 0 && s[k] != s[i] {
            k = π[k-1]
        }
        if s[k] == s[i] {
            k++
        }
        if i < len(p) {
            π[i] = k
        }
        πPrev = k
        if k == len(p) {
            result = append(result, i-2*len(p))
        }
    }
    return result
}

File: sprint8/By languages/C#/lesson6code1.cs
public static List<int> Search(string p, string text)
{
    List<int> result = new List<int>();
    string s = p + "#" + text;
    int[] π = new int[p.Length];
    int π_prev = 0;
    for (int i = 1; i <= s.Length; i++)
    {
        int k = π_prev;
        while (k > 0 && s[k - 1] != s[i - 1])
        {
            k = π[k - 1];
        }
        if (s[k - 1] == s[i - 1])
        {
            k++;
        }
        if (i <= p.Length)
        {
            π[i - 1] = k;
        }
        π_prev = k;
        if (k == p.Length)
        {
            result.Add(i - 2 * p.Length);
        }
    }
    return result;
}
